Add support for table_constraints of AdbcConnectionGetObjects()

kou opened the pull request:
See
https://arrow.apache.org/adbc/19/cpp/api/group__adbc-connection-metadata.html#ga32f64f4cfc86140e2d8d02b328215dbf for `AdbcConnectionGetObjects()` and `table_constraints` specification.

kou committed:
Add support for table_constraints of AdbcConnectionGetObjects()

See
https://arrow.apache.org/adbc/19/cpp/api/group__adbc-connection-metadata.html#ga32f64f4cfc86140e2d8d02b328215dbf
for `AdbcConnectionGetObjects()` and `table_constraints`
specification.

pdet reviewed (changes requested):
Many thanks for the PR!

Could you maybe add a separate test where we verify different constraint combinations?

For example: unique, check, and foreign_keys.

Also try different orders of column constraints, as well as both single-column and multi-column constraints.

Basically, we should verify that the options are all properly set in `CONSTRAINT_SCHEMA` and `USAGE_SCHEMA` (in the case of foreign keys).

kou committed:
Add more tests

kou commented:
> Could you maybe add a separate test where we verify different constraint combinations?
> 
> For example: unique, check, and foreign_keys.

Sure. I've added (single column) `PRIMARY KEY`, `NOT NULL`, `CHECK`, `UNIQUE` and `REFERENCES` cases because https://duckdb.org/docs/stable/sql/constraints.html lists them.

Some notes:
* `NOT NULL` isn't returned because `information_schema.constraint_column_usage` doesn't include it.
* `REFERENCES` (foreign key) doesn't return foreign table and referenced key information as `USAGE_SCHEMA` because `information_schema` https://duckdb.org/docs/stable/sql/meta/information_schema.html doesn't provide foreign table and referenced key information. Can we make it out-of-scope of this PR? We can revisit this once `information_schema` returns the information.
  * FYI: `information_schema.constraint_column_usage` returns `FOREIGN KEY (foreign_id) REFERENCES foreign_table(id)` as a text but I don't want to parse it...

> Also try different orders of column constraints, as well as both single-column and multi-column constraints.

Could you clarify "different orders of column constraints"? What case do you imagine?

```sql
CREATE TABLE my_table (
  primary_key INTEGER PRIMARY KEY,
  custom_check INTEGER CHECK (custom_check > 0)
);
```

and


```sql
CREATE TABLE my_table (
  custom_check INTEGER CHECK (custom_check > 0),
  primary_key INTEGER PRIMARY KEY
);
```

?

Or

```sql
CREATE TABLE my_table (
  a INTEGER,
  b INTEGER,
  PRIMARY KEY (a, b)
);
```

and

```sql
CREATE TABLE my_table (
  a INTEGER,
  b INTEGER,
  PRIMARY KEY (b, a)
);
```

?

> Basically, we should verify that the options are all properly set in `CONSTRAINT_SCHEMA` and `USAGE_SCHEMA` (in the case of foreign keys).

Can we make `USAGE_SCHEMA` support out-of-scope of this PR? (See the above foreign key note.)

pdet commented:
Hi @kou, thanks for the quick changes.

I had 
```
CREATE TABLE my_table (
  a INTEGER,
  b INTEGER,
  PRIMARY KEY (b, a)
);
```
 in mind, but if you think the other is also interesting please also add a test to it.

I think it is fine to have `USAGE_SCHEMA` outside the scope of this PR.

kou committed:
Add different order multi-column UNIQUE index case

kou commented:
Thanks for clarifying it.

I've added a test for the case. But I used `UNIQUE` not `PRIMARY KEY` because we can't add multiple `PRIMARY KEY` to the same table:

```sql
CREATE TABLE my_table (
  a INTEGER,
  b INTEGER,
  UNIQUE (b, a)
);
```

BTW, the current implementation (the following SQL) works well for different order:

```sql

				column_constraints AS (
					SELECT
						table_catalog,
						table_schema,
						table_name,
						constraint_name,
						constraint_type,
						LIST(column_name) as column_names
					FROM information_schema.constraint_column_usage
					WHERE column_name LIKE '%s'
					GROUP BY
						table_catalog,
						table_schema,
						table_name,
						constraint_name,
						constraint_type
				)
```

But I feel that it depends on the current implementation. I think that we must use `LIST(column_name ORDER BY ...) as column_names` to ensure producing the correct order. But `information_schema.constraint_column_usage` doesn't have information for the `ORDER BY`.

Can we defer it to a separated PR like `USAGE_SCHEMA`?

The other case is covered by already added test because it uses the following that includes multiple constraints in one table:

```sql
			CREATE TABLE my_table (
				primary_key INTEGER PRIMARY KEY,
				not_null INTEGER NOT NULL,
				custom_check INTEGER CHECK (custom_check > 0),
				unique_single INTEGER UNIQUE,
				unique_multi1 INTEGER,
				unique_multi2 INTEGER,
				UNIQUE (unique_multi2, unique_multi1),
				foreign_id INTEGER REFERENCES foreign_table(id)
			)
```

> I think it is fine to have `USAGE_SCHEMA` outside the scope of this PR.

Thanks. I've updated the PR description that mentions that this PR doesn't add support for `USAGE_SCHEMA`.

kou committed:
Use duckdb_constraints()

kou reviewed (commented):
OK. I've changed to use `duckdb_constraints()`. Now, foreign key is supported.

kou commented:
@pdet Could you take a look at the `duckdb_constraints()` approach?

pdet reviewed (approved):
Thanks for the ping!
I think it's looking excellent, thanks for all the effort you put in the PR and in the changes!

kou commented:
@Mytherin Could you merge this approved PR?

Mytherin commented:
Thanks!
